Add a dimensionality reduction supplemental notebook (#24)
AlexsLemonade/2020-july-training: supplemental_notebooks/2020-07-29_dimensionality_reduction.Rmd

Perhaps unsurprisingly, this can be very hard to visualize, as our brains have trouble thinking in more than 3 dimensions, and paper or computer screens effectively only show two at a time (maybe 3 with color and color vision).
We have some techniques that are useful, such as heatmaps of the gene expression matrix, but these can be hard to interpret, and it is extremely hard to accurately show relationships among samples in a grid (hierarchical trees are useful, but also confusing and easily misinterpreted).

So what we would like to do is take this information with all of its dimensions and reduce it down to something we can visualize, while maintaing as much of the underlying information as possible. 

### Penguins

To illustrate this, we will use a fun little data set called _Palmer Penguins_. This is contained in an R package may not be installed on your machine, so you can install it for your user with:

```{r}
install.packages("palmerpenguins")
```

We can then load the dataframe into our environment and look at its structure with:

```{r}
penguins <- palmerpenguins::penguins
str(penguins)
```

What we have is a data frame (`tibble`, really) with 8 columns, so 8 dimensions!

> There is some missing data in there, so lets clean that up:

Reviewer: ```suggestion
There is some missing data in there, so let's clean that up:
```

```{r}

> penguins <- drop_na(penguins)

Reviewer: I might add a comment about what this is doing. I don't remember what `drop_na()` does (same as `complete.cases()`? ) so I would appreciate the context, but maybe you don't want to get into it!

Author: Yeah, this way it is the same as `filter(complete.cases(.))`. It can also be applied to a single column: `drop_na(species)`

We'll start with an interactive 3d plot, using a package called `plotly`. 
The main function `plot_ly()` works much like `ggplot`, but with a somewhat different format for its arguments, and using pipes (`%>%`) instead of plusses (`+`)

```{r}
plot_ly(penguins,
        x = ~ bill_length_mm,
        y = ~ bill_depth_mm,
        z = ~ flipper_length_mm,
        color = ~ body_mass_g) %>%
  add_markers(size = 0.5)
```

You can drag around the 3d plot there to get different views, but it can be hard to find one that shows all of the structure in the data from a single view.
Nonetheless, it should be clear that there are some views that are better than others, and if we display the data from those perspectives, it will show more of the variation in the data set than others where all the points are stacked on top of one another

This is where dimensionality reduction will help out!

### Throwing out data

The simplest from of dimensionality reduction is to just throw out a dimension.
This is what we did when we selected only the most variable genes. 
As you might guess, this can be fraught, but it is also what happens any time we make a plot of just two dimensions of our data.

> Below is a plot of just the bill length and bill depth 
> There is nothing wrong with this plot, but it is certainly not a view of the full data set, and it probably doesn't show the same clarity of clustering that you might have been able to infer from playing with the 3D plot

Reviewer: ```suggestion
Below is a plot of just the bill length and bill depth.
There is nothing wrong with this plot, but it is certainly not a view of the full data set, and it probably doesn't show the same clarity of clustering that you might have been able to infer from playing with the 3D plot.
```

Now let's move up to three dimensions, adding in flipper length!
```{r}
penguin_lengths <- penguins[,3:5] 

penguin_length_pca <- prcomp(penguin_lengths, scale = TRUE)
```

```{r}
ggplot(as.data.frame(penguin_length_pca$x), 
       aes(x = PC1, y = PC2,
           color = penguins$species)) +
  geom_point() + 
  colorblindr::scale_color_OkabeIto()
```

It turns out that this looks quite similar overall, though some points may have moved. 
This is because it turns out that there are some reasonable correlations between dimensions in the data, and in particular between bill width and flipper length:

```{r}
cor(penguins$bill_length_mm, penguins$flipper_length_mm)
```

PCA is in effect rotating the data to remove correlations between the new axes, and correlated data end up collapsed into a single dimension.

If you play around with the 3D plot, you should be able to find a rotation that gives a similar view to the plot above.

Note that this plot does not necessarily give the best separation of species!
The PCA we performed was blind to that information, and only finds a projection based on variation, not species id.

> We can still plot the remaing dimension, just to see if it retains information we care about:

Reviewer: ```suggestion
We can still plot the remaining dimension, just to see if it retains information we care about:
```

```{r}
ggplot(as.data.frame(penguin_length_pca$x), 
       aes(x = PC3,
           fill = penguins$species)) +
  geom_histogram() + 
  colorblindr::scale_fill_OkabeIto()
```

There may still be some information there, especially distinguishing Adelie and Chinstrap penguins, but it is likely less than we get from the first two dimensions, so we may be comfortable discarding it.

> ### What about UMAP & tSNE?

Reviewer: ```suggestion
### What about UMAP & t-SNE?
```

> UMAP and tSNE allow non-linear transforamations of the dimensions, but perform the conceptually similar task of creating a transformation that captures most of the variation in the multidimenstional data in two axes.

Reviewer: ```suggestion
UMAP and t-SNE allow non-linear transformations of the dimensions, but perform the conceptually similar task of creating a transformation that captures most of the variation in the multidimensional data in two axes.
```

Let's see what the UMAP projection looks like:
```{r}
penguin_length_umap <- uwot::umap(penguin_lengths)

ggplot(as.data.frame(penguin_length_umap),
       aes(x = V1, y = V2, 
       color = penguins$species)) + 
  geom_point() + 
  colorblindr::scale_color_OkabeIto()
```

You can see that this has separated the data into much more discrete clusters (that do not always align with species), but that the data no longer look much at all like the original data. 

In this data, the results are also unstable: repeating the above code gives a different looking result!

Let's see what the UMAP projection looks like:
```{r}
penguin_length_umap <- uwot::umap(penguin_lengths)

ggplot(as.data.frame(penguin_length_umap),
       aes(x = V1, y = V2, 
       color = penguins$species)) + 
  geom_point() + 
  colorblindr::scale_color_OkabeIto()
```

> What about tsne? Let's have a look:

Reviewer: ```suggestion
What about t-SNE? Let's have a look:
```
